### PR TITLE
PMM-15406 Test the ServiceNow delivery connection

### PR DESCRIPTION
**Pull request by nachodd**
## What

Adds a **Test connection** action to the ServiceNow connection settings tab, and normalizes the delivery endpoint before it is stored.

Before this, the tab could save a receiver and a credential but never say whether either worked. The first evidence of a wrong endpoint or a bad credential was a failed send — after a collection had already run, and phrased in the receiver's own words. That was observed during the Tech Preview validation pass (PMM-15381): an operator saved an endpoint and a credential, the tab accepted both without objection, and the misconfiguration surfaced only when a later send failed at `resolve_ticket` with a receiver-side 400 that named nothing the operator had chosen.

Base branch: `PMM-15419-servicenow-credential-copy` (which targets `PMM-15205-sep-fb`). This is the ticket that makes that branch's **"Verify and connect"** button actually verify.

## Feature build

- Image: `perconalab/pmm-server-fb:PR-4560-<short-sha>` once the build is green — [Percona-Lab/pmm-submodules#4560](https://github.com/Percona-Lab/pmm-submodules/pull/4560) (draft; Jenkins link is in that PR's checks).
- To exercise the probe, the SEP side-car must carry SEP-1941. `SEP/sidecar/pmm-fb/compose.yaml` pins `percona/percona-sep:249f6738…`, which predates it and does not accept `delivery` as a target; `b97ee985fc64f841f611b3a057737814da308a61` (2026-09-03) does.
- The branch is stacked, so the image carries `PMM-15205-sep-fb` and PMM-15419 beneath this ticket — the whole ServiceNow connection tab, not only the test action.

## How

**The action.** It lives on the connected screen, beside *Renew credentials* and *Disconnect*, and it saves nothing. The probe reads SEP's own stored delivery inputs server-side, so an operator can ask whether what is configured works without first replacing it — and typed-but-unsaved credentials are not something SEP can probe with, which is why the form does not offer it. It also runs once, on its own, right after a successful save.

**Two failures kept apart.** A rejected call — 401, 403, a network failure, a timeout on PMM's side — means the probe never ran, and is reported as that. A `ConnectivityResult` that came back means it did run, and is reported by its own status, whatever that status says.

**Every outcome has its own label.** `Settings.messages.ts` carries one operator-facing sentence per member of SEP's `ConnectivityStatusEnum`, and both that record and the severity map are keyed by the generated union — a member added to SEP fails to compile here rather than rendering an empty alert. Until one is written, an unrecognised status falls back on the result's `reachable` boolean. `not_configured` and `probe_undeclared` are informational, not errors: `probe_undeclared` in particular reads as "nothing to test here", since a shipped side-car whose plan declares no probe step is a realistic answer rather than an edge case.

**SEP's `detail` is deliberately not rendered.** For the delivery target it is always a fixed English sentence (`_probe_delivery` passes no `detail` on the exception path, so `_DEFAULT_DETAILS` applies), except for the two unavailable outcomes where it is the resolver's prose. Our copy says both better, and showing it would put back exactly the "failure phrased in the receiver's own words" this ticket set out to remove.

**Endpoint normalization.** `buildDeliveryInputsPatch` now stores the endpoint as SEP will use it. SEP's `split_endpoint` keeps only `scheme://netloc` as the transport origin and moves the path onto the plan's step paths, joining each with the endpoint path's trailing slash stripped — so `https://host//` is reduced inside SEP while the tab went on displaying it verbatim. Normalization removes only what SEP provably discards (trailing slashes, the fragment) and rewrites nothing else:

- the query is preserved exactly as typed rather than re-serialized — SEP's `parse_qsl` collapses a repeated key and drops a blank value on its own, and matching that here would delete typed input to no visible gain;
- a URL carrying userinfo is left untouched, because `URL.origin` omits `user:pass@` where Python's `netloc` keeps it, so normalizing one would store an endpoint that authenticates differently from the one SEP receives;
- a blank endpoint stays blank — it still means "keep the receiver this image bakes in".

**A verdict never outlives what it described.** It is true of the configuration stored when the probe ran, so a configuration that changes underneath it — saved from another session, arriving on a refetch — clears it rather than leaving it standing beside a receiver it never reached. It is also withheld while a re-test is in flight, since React Query keeps the previous `data` through a pending mutation. One limitation is inherent: SEP masks every stored secret with the same string, so a same-endpoint credential change made elsewhere is not detectable from the settings response.

**The vendored client.** Only `useConnectivityCheck` was brought across from SEP (verbatim, reformatted), plus its exports. `specs/sep.json` and `generated/sep.ts` already carried SEP-1941 — `ServiceEnum` has `delivery` and `ConnectivityStatusEnum` has all nine members — having arrived with the SEP sync in #5886, so nothing was regenerated.

## Acceptance criteria

- [x] Test connection action probes the receiver and reports the outcome inline, without leaving the tab.
- [x] Every `ConnectivityStatusEnum` member has its own label; a later addition falls back to `reachable`.
- [x] A probe that never ran is distinguished from a probe that ran and found the connection bad.
- [x] Testing does not save and never mutates the stored configuration.
- [x] Admin-only — the settings page renders nothing to a non-admin (`Page roles={[OrgRole.Admin]}`), and SEP holds the endpoint to `IsApiAdmin`.
- [x] No secret reaches the form, the result, or any message the test produces.
- [x] The endpoint is normalized before it is stored; a blank one is never turned into a stored one.
- [x] The test result is its own outcome and cannot contradict the tab's derived connection status.
- [x] The delivery target and its outcome states come from `@sep/api` generated types (`ConnectivityCheckRequest['targets']`, `ConnectivityStatus`), not a hand-written enum.

## Out of scope

- The probe itself — SEP-1941, tracked on this side by PMM-15418.
- What the form submits, settled by PMM-15294.
- Whether the receiver accepts a given bundle size, and attributing a failure on an already-started send.

## Known boundary

A deployment whose baked delivery plan declares no probe step answers `probe_undeclared` rather than a verdict, and whether the production ServiceNow credential can read a probe target is still unverified — SEP-1941 gated the baked side-car probe step on exactly that question. The label is written for it.

## Test plan

- `cd ui/apps/pmm && npx vitest run` — 524 passed, 13 skipped.
- `cd ui/packages/sep/api && npx vitest run` — 106 passed; `packages/sep/framework` — 748 passed.
- `npx tsc --noEmit` clean in both `apps/pmm` and `packages/sep/api`; `oxlint` and `oxfmt --check` clean.
**Live verification — done**, against a running PMM (3.8.0) + SEP on the dev server, signed in as an admin:

- The tab renders the connected screen with the baked receiver (`https://percona.service-now.com`) and offers **Test connection**.
- The action posts exactly `{"targets":["delivery"]}` to `/sep/api/sep/admin/connectivity-check/` — one target, captured off the wire.
- This SEP image bakes no probe step, so the verdict is the `probe_undeclared` label: *"This PMM version doesn't ship a connectivity test for ServiceNow, so there's nothing to check. The saved connection is unaffected."* That is the boundary named under **Known boundary** above, reached in practice rather than in theory.
- Opening a stored connection fires no probe; only a save does.
- Forcing the probe call to 403 renders the error alert alone, with no result alert, and leaves the stored endpoint untouched — criterion 3, end to end.

That last case found a real bug, fixed in a49f3d4ad: a refused probe reported *"Your account isn't allowed to change this connection. A PMM administrator has to save it."* — `sepErrorMessage`'s wording, written for the PATCH. A 403 on a test says nothing about saving. `probeErrorMessage` now gives the test its own 401 / 403 / network copy and skips the settings-422 lookup, and a unit test asserts the message is *not* the save-path string.

Not covered live: the save → auto-probe path and the drifted state, both of which would mean writing credentials to the environment; they stay on unit coverage. The screenshots below come from a headless engine whose text layout differs from Chromium, so treat button wrapping in them as an artifact rather than a styling defect.

### CODE REVIEW

File: ui/apps/pmm/src/pages/settings/Settings.messages.ts [hunk 1]
```diff
@@ -167,6 +168,39 @@ export const Messages = {
...
+        timeout: "ServiceNow didn't answer in time. Try again.",
+        not_configured:
+          "There's nothing to test yet — connect ServiceNow first.",
+        inputs_drifted:
+          'The saved credentials no longer match what this PMM version expects. Renew them to reconnect.',
+        // Not an error and not the operator's doing: this image's delivery plan
+        // declares no probe, so there is nothing to reach out to.
+        probe_undeclared:
+          "This PMM version doesn't ship a connectivity test for ServiceNow, so there's nothing to check. The saved connection is unaffected.",
+      } as Record<ConnectivityStatus, string>,
```
[the remark this answers is not in the document]
Comment (nachodd) [reply ui/apps/pmm/src/pages/settings/Settings.messages.ts]: Good catch — fixed in de0386264. `as` is an assertion, so it silenced a missing member rather than reporting one; the comment claimed a check the code did not perform. Now `satisfies Record<ConnectivityStatus, string>`, verified by deleting `timeout` locally: `TS1360: Type '{...}' does not satisfy the expected type 'Record<...>'` at the definition, where the `as` version compiled clean.

File: ui/apps/pmm/src/pages/settings/components/servicenow/ServiceNowConnection.utils.ts [hunk 2]
```diff
@@ -105,20 +109,75 @@ export const toFormValues = (
...
+    return trimmed;
+  }
+  if (url.protocol !== 'http:' && url.protocol !== 'https:') {
+    return trimmed;
+  }
+  if (url.username || url.password) {
+    return trimmed;
+  }
+  return `${url.origin}${url.pathname.replace(/\/+$/, '')}${url.search}`;
+};
```
[the remark this answers is not in the document]
Comment (nachodd) [reply ui/apps/pmm/src/pages/settings/components/servicenow/ServiceNowConnection.utils.ts:165]: Half right, and the half that is right was my wording — fixed in de0386264.

The query is not re-serialized: `url.search` passes existing percent-encoding through untouched, case included (`?a=%2f&b=e+f` and `?a=b%20c` all come back byte-identical). The only change it makes is escaping a character that could not have stood in a URL unescaped — a literal space becomes `%20` — which decodes identically under SEP's `parse_qsl`. So slicing the raw `?…` out of the input string would preserve nothing extra that reaches the receiver, at the cost of hand-rolling URL parsing.

What was wrong was the docstring saying the query is kept "exactly as typed" while also claiming the `URL` constructor normalizes encoding case. It doesn't. The comment now states both behaviours accurately.

**Comment by yyyyyyyan** [comment]
@copilot resolve the merge conflicts in this pull request

**Review by yyyyyyyan — APPROVED** [review]
